### PR TITLE
Add grammar in docs for {f32,f64}::from_str, mention known bug.

### PR DESCRIPTION
**Pull request by frewsxcv**
- Original bug about documenting grammar
  - https://github.com/rust-lang/rust/issues/32243
- Known bug with parsing
  - https://github.com/rust-lang/rust/issues/31407

### CODE REVIEW

File: src/libcore/num/dec2flt/mod.rs [hunk 1]
```diff
@@ -122,11 +122,35 @@ macro_rules! from_str_float_impl {
...
+            /// (\+|-)?
+            /// (inf|
+            ///  NaN|
+            ///  ([0-9]+|
+            ///   [0-9]+\.[0-9]*|
+            ///   [0-9]*\.[0-9]+)
+            ///  ((e|E)
+            ///   (\+|-)?
+            ///   [0-9]+)?)
+            /// ```
```
Comment (frewsxcv) [inline src/libcore/num/dec2flt/mod.rs]: I'm not tied to the particular formatting of the regular expression here. You can visualize this [as a railroad diagram here](https://regexper.com/#%28%5C%2B%7C-%29%3F%28inf%7CNaN%7C%28%5B0-9%5D%2B%7C%5B0-9%5D%2B%5C.%5B0-9%5D*%7C%5B0-9%5D*%5C.%5B0-9%5D%2B%29%28%28e%7CE%29%28%5C%2B%7C-%29%3F%5B0-9%5D%2B%29%3F%29).

Comment (frewsxcv) [reply src/libcore/num/dec2flt/mod.rs]: Also if people would prefer, I can migrate this to BNF, if people think that'd be easier for readers

File: src/libcore/num/dec2flt/mod.rs [hunk 2]
```diff
@@ -122,11 +122,35 @@ macro_rules! from_str_float_impl {
...
+            ///
+            /// # Known bugs
```
Comment (frewsxcv) [inline src/libcore/num/dec2flt/mod.rs]: Historically, we haven't mentioned known bugs in the docs, but considering this known bug has existed for years and people keep hitting it, I think it's worth calling it out here.

File: src/libcore/num/dec2flt/mod.rs (hunk 1 above)
Comment (nagisa) [reply src/libcore/num/dec2flt/mod.rs]: We could embed SVG of the railroad in-line… would look nice but barely maintainable.

Comment (nagisa) [reply src/libcore/num/dec2flt/mod.rs]: (Rust has [a library do draw these diagrams](https://crates.io/crates/railroad) too)

**Comment by alexcrichton** [comment]
FWIW this isn't the [first instance of documenting a known issue](https://doc.rust-lang.org/1.30.0/std/sync/mpsc/struct.Receiver.html#known-issues), and I think while we don't really have a policy around this the defacto policy is that "if someone cares enough to document the issue while not being willing to fix it, then seems fine to document".

**Comment by QuietMisdreavus** [comment]
I'm not a fan of including a grammar or regular expression like that, especially in the std docs. We've gradually been moving any kind of grammar into the Reference, so i would consider that a much better place for that.

As far as documenting known issues, i'm okay with it, but i think it could be better phrased:

```markdown
# Known issues

In some situations, some strings that should create a valid float instead return an
error. See [issue #31407] for details.
```

**Comment by QuietMisdreavus** [comment]
> do you have any examples of sections in the Reference that document specific standard library APIs?

There's a ["Special types and traits"](https://doc.rust-lang.org/nightly/reference/special-types-and-traits.html) page that seems to document some of the more-common items from the standard library.

But i think i get your point. Since this is library code, it makes sense to document it here. (Though i'm not sure if we ever got together a better documentation of *all* type conversion functionality than the `std::convert` module docs, and that doesn't even include the `FromStr` trait being documented here...)

There is [a grammar for floating-point literals](https://doc.rust-lang.org/nightly/reference/tokens.html#floating-point-literals) in the reference; does this match the regular expression you've entered there? I guess it wouldn't include the type suffix, but i think everything else matches?

**Comment by frewsxcv** [comment]
>  but i think everything else matches?

A couple other differences come to mind:

- float literals, like integer literals, can have underscores
- float literals require a decimal point
- float literals can't have a leading `+`

**Comment by frewsxcv** [comment]
Also looks like float literals can have a capital `E` in the exponent part

**Comment by QuietMisdreavus** [comment]
cc @rust-lang/wg-grammar - I think some people in there aren't in the docs group on github.

**Comment by Centril** [comment]
@QuietMisdreavus I'm not sure what the ping (wg-grammar) was for; can you elaborate?

**Comment by QuietMisdreavus** [comment]
@Centril We talked about it on Discord, but my question was effectively "does the grammar WG care about the grammar of library functions", which turns out to be "not officially".

@frewsxcv Coming back to this, it's fine to have it here, but i find the regex a little hard to read. Can you convert it to BNF? I think that's a little easier to decode at a glance if you're not familiar with the syntax.

**Comment by frewsxcv** [comment]
Here's my attempt at writing out the grammar using [W3C's EBNF syntax](https://www.w3.org/TR/REC-xml/#sec-notation):

```
Float  ::= Sign? ( 'inf' | 'NaN' | Number )
Number ::= ( Digit+ |
             Digit+ '.' Digit* |
             Digit* '.' Digit+ ) Exp?
Exp    ::= [eE] Sign? Digit+
Sign   ::= [+-]
Digit  ::= [0-9]
```

It can be visualized [here](http://www.bottlecaps.de/rr/ui). Thoughts? Is this more clear than the regex?

File: src/libcore/num/dec2flt/mod.rs [hunk 3]
```diff
@@ -122,11 +122,33 @@ macro_rules! from_str_float_impl {
...
             /// * '5.'
-            /// * '.5', or, equivalently,  '0.5'
+            /// * '.5', or, equivalently, '0.5'
             /// * 'inf', '-inf', 'NaN'
             ///
             /// Leading and trailing whitespace represent an error.
             ///
+            /// # Grammar
+            ///
+            /// All strings that adhere to the following EBNF grammar
```
Comment (Centril) [inline src/libcore/num/dec2flt/mod.rs]:
```suggestion
            /// [EBNF]: https://www.w3.org/TR/REC-xml/#sec-notation
            ///
            /// All strings that adhere to the following [EBNF] grammar
```

**Comment by Centril** [comment]
Picking someone from T-libs randomly...

r? @Kimundi

**Comment by jkordish** [comment]
ping from triage. @Kimundi have you had time to review?

File: src/libcore/num/dec2flt/mod.rs [hunk 4]
```diff
@@ -122,11 +122,35 @@ macro_rules! from_str_float_impl {
...
             /// * '5.'
-            /// * '.5', or, equivalently,  '0.5'
+            /// * '.5', or, equivalently, '0.5'
             /// * 'inf', '-inf', 'NaN'
             ///
             /// Leading and trailing whitespace represent an error.
             ///
+            /// # Grammar
+            ///
+            /// [EBNF]: https://www.w3.org/TR/REC-xml/#sec-notation
```
Comment (QuietMisdreavus) [inline src/libcore/num/dec2flt/mod.rs]: Nit: Don't we usually put link anchors underneath their first use?

**Review by QuietMisdreavus — APPROVED** [review]

**Pull request merged**